packages/grafbase-sdk/src/rateLimiting.ts

```diff
@@ -0,0 +1,162 @@
+}
+
+export interface IpCondition {
+  ips: string[] | RateLimitingAllowAny
+}
+
+export interface JwtCondition {
+  jwtClaims: JwtConditionSpec[]
+}
+
+export interface JwtConditionSpec {
+  name: string
+  value: any | RateLimitingAllowAny
+}
+
+export interface RateLimitingRule {
+  name: string
+  limit: number
+  duration: number
+  condition: RateLimitingRuleCondition
+}
+
+type RateLimitingRuleCondition =
+  | HeaderCondition
+  | OperationCondition
+  | JwtCondition
+  | IpCondition
+
+export interface RateLimitingParams {
+  rules: RateLimitingRule[]
+}
+
+export class RateLimiting {
+  rules: RateLimitingRule[]
+
+  constructor(rateLimitingParams: RateLimitingParams) {
+    this.rules = rateLimitingParams.rules
+  }
+
+  public toString(): string {
+    return `\nextend schema\n  @rateLimiting(rules: ${renderRules(this.rules)}\n  )\n`
+  }
+}
+
+function renderRules(rules: RateLimitingRule[]): string {
+  const renderedRules = rules.map((rule) => {
+    const name = `\n      name: "${rule.name}"`
+    const limit = `,\n      limit: ${rule.limit}`
+    const duration = `,\n      duration: ${rule.duration}`
+
+    const condition = `,\n      condition:${renderCondition(rule.condition)}`
+
+    return `{${name}${limit}${duration}${condition}\n    }`
+  })
+
+  return `[${renderedRules}]`
+}
+
+function renderCondition(condition: RateLimitingRuleCondition): string {
+  if (isHeaderCondition(condition)) {
+    return renderHeaderCondition(condition)
+  }
+  if (isJwtCondition(condition)) {
+    return renderJwtCondition(condition)
+  }
+  if (isIpCondition(condition)) {
+    return renderIpCondition(condition)
+  }
+  if (isOperationCondition(condition)) {
+    return renderOperationCondition(condition)
+  }
+  return ''
+}
+
+function isHeaderCondition(
+  condition: RateLimitingRuleCondition
+): condition is HeaderCondition {
+  return (condition as HeaderCondition).headers !== undefined
+}
+
+function renderHeaderCondition(condition: HeaderCondition): string {
+  const headerConditions = condition.headers
+    .map((headerCondition) => {
+      let value = '"*"'
+      if (headerCondition.value !== '*') {
+        value = `[${headerCondition.value.map((value) => `"${value}"`).join(', ')}]`
+      }
+
+      return `{name: "${headerCondition.name}", value: ${value}}`
+    })
+    .join(',')
+
+  return ` {\n        headers: [${headerConditions}]\n      }`
+}
+
+function isJwtCondition(
+  condition: RateLimitingRuleCondition
+): condition is JwtCondition {
+  return (condition as JwtCondition).jwtClaims !== undefined
+}
+
+function renderJwtCondition(condition: JwtCondition): string {
+  const conditions = condition.jwtClaims
+    .map((condition) => {
+      let value = '"*"'
+      if (condition.value !== '*') {
+        if (typeof condition.value === 'object') {
+          value = `${JSON.stringify(JSON.stringify(condition.value))}`
+        } else {
+          value = `"${condition.value}"`
+        }
+      }
```

Comment:
Any reason not to always do `JSON.stringify` like you're doing on the object path?  Seems like it'd be more robust to just always do it, although I may be missing something.